### DISCUSSION TITLE
Guidance on nicegui coding patterns

### DISCUSSION THREAD

**Discussion by BlankAdventure** [zauberzeug/nicegui#2939] [Q&A] [answered]
### Question

Okay, this is a bit of a long one, but not actually that complicated. Currently, I have multiple projects/repos/apps that all use nicegui. I want to put together a personal/showcase site that allows (multiple, independent) users to play with these apps live. However, I still want the individual projects/apps to be able to be run in a standalone manner, which I find helpful for debugging/development, and for such case as when someone wants to use that specific repo/app.

My showcase code would look something like this -

```python
from nicegui import ui


@ui.page('/appA') #Launch app A
def appA():
    pass
    
@ui.page('/appB') #Launch app B
def appB():
    pass

@ui.page('/appC') #Launch app C
def appC():
    pass
    
 
ui.label('My Showcase Site').classes('text-xl')

ui.link('[App A]', '/appA')
ui.link('[App B]', '/appB')
ui.link('[App C]', '/appC')


ui.run()

```

Then, I figure there are 3 possible "patterns" for writing the individual apps - 

1. (A) Global Pattern:

```python
# Global Pattern 

from nicegui import ui
import numpy as np

def update():
    label.set_text( np.random.randint(0,10,1) )

with ui.row():
    ui.button('Click (A)' ,on_click=update)
    label = ui.label()
        
    
if __name__ in {"__main__", "__mp_main__"}:    
    ui.run()
```

This is a very common pattern for nicegui code. We include the guard to code so that it can be run in standalone execution, but not automatically if imported.

2. (B) Inner-Function Pattern:

```python
# Inner-Function Pattern

from nicegui import ui
import numpy as np

def main():
    
    def update():
        label.set_text( np.random.randint(0,10,1) )
    
    with ui.row():
        ui.button('Click (B)' ,on_click=update)
        label = ui.label()
        
    
if __name__ in {"__main__", "__mp_main__"}:    
    main()
    ui.run()
        
```

This again is a common nicegui pattern. We use the guard to permit running the code in standalone execution, but not when imported.

3. (C) Class Pattern:

```python
# Class pattern

from nicegui import ui
import numpy as np

class App():
    def __init__(self):
        self.setup()

    def setup(self):
        with ui.row():
            ui.button('Click (C)' ,on_click=self.update)
            self.label = ui.label()
    
    def update(self):
        self.label.set_text( np.random.randint(0,10,1) )

        
    
if __name__ in {"__main__", "__mp_main__"}:    
    app = App()
    ui.run()
```

Here, we encapsulate nicegui into a class.  Again, we use the guard to permit running the code in standalone execution, but not when imported.

Returning to the showcase site, we build it it out as follows:

```python
from nicegui import ui


@ui.page('/appA') #Global pattern -> This doesn't work 
def appA():
    import A
    
@ui.page('/appB') #Inner-function pattern -> this works
def appB():
    from B import main
    main()

@ui.page('/appC') #Class pattern -> this works
def appC():
    from C import App
    App()

    
ui.label('My Showcase Site').classes('text-xl')
ui.link('[App A]', '/appA')
ui.link('[App B]', '/appB')
ui.link('[App C]', '/appC')
ui.run()
```

So now, we find that the Global Pattern doesn't work for us - it cannot be imported into 'this' page and made available to multiple users in a unique instance. Now that's probably *fine* because the inner-function pattern and class patterns *do* solve my problem (apps can be developed and run standalone, or, launched from the showcase site). Next, I don't really like the inner function pattern as IMO, it makes debugging and testing more challenging. So that leaves us with the class pattern and what my question ultimately boils down to: this solves my described problem, but I very rarely see this strategy used in the nicegui community and so I can't help but wonder if there's some downside that I'm missing here?

Just looking for thoughts and feedback, thanks!

(edit: what's interesting is when I first started playing with nicegui a year or so ago, I class-ified everything, coming from a background in traditional UI development. I shifted away from this as none of the coding examples did it this way)

**Comment by rodja** [comment, marked as answer]
There is nothing wrong with the class-pattern you suggested. We use it intensively in [RoSys, our framework for mobile robotics](https://rosys.io) (see our [videos page](https://github.com/zauberzeug/rosys/blob/main/rosys/analysis/videos_page_.py) as just one example). But you are right. It was missing in the modularization example. I just created #2944 to fix that. Most other examples in the docs are focusing on other aspects and therefor use the simplest possible strategy (eg. "(A) Global Pattern"). This makes it very much over-represented in the docs. But the alternative would mean a lot of "unnecessary" code.

**Reply by BlankAdventure** [reply]
Thanks, that makes sense. I saw the new modularization examples as well.
